src/components/core/GeoChart.tsx

```diff
@@ -70,39 +103,136 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+        this.validateData(executionResult, geoData, limit, onDataTooLarge);
+        const { data } = executionResult;
+        // transformation data
+        const { size, segmentBy } = geoData;
+        const attributeHeaderItems = getGeoAttributeHeaderItems(executionResult, geoData);
+        let series;
+        let sizeValues: number[] = [];
+        if (size && isTwoDimensionsData(data)) {
+            sizeValues = data[size.index].map(stringToFloat);
+        }
+        if (segmentBy) {
+            const segmentByData = attributeHeaderItems[segmentBy.index];
+            series = this.getSegmentBySeries(segmentByData);
+            this.setState({
+                legendIndexEnabled: new Array(series.length).fill(true),
+                legendNamesEnabled: [...series],
+            });
+        }
+
+        this.geoChartOptions = {
+            geoData,
+            series,
+            sizeValues,
+        };
+    }
+
     private renderChart = (): React.ReactElement => {
         const { config, chartRenderer, execution } = this.props;
+        const segmentByIndex: number = get(this.geoChartOptions, "geoData.segmentBy.index");
         const chartProps = {
-            config,
+            config:
+                segmentByIndex !== undefined
+                    ? { ...config, selectedSegmentItems: [...this.state.legendNamesEnabled] }
+                    : config,
             execution,
         };
         return chartRenderer(chartProps);
     };
 
     private renderLegend = (): React.ReactElement => {
-        const { config, execution, legendRenderer, locale, legendPostion: position } = this.props;
-        const legendProps = {
+        const {
+            config,
+            legendRenderer,
+            locale,
+            legendPosition: position,
+            execution: { executionResponse },
+        } = this.props;
+
+        let legendProps = {
             config,
-            execution,
             locale,
             position,
+            geoData: {},
         };
+        if (this.geoChartOptions) {
+            const { size, segmentBy } = this.geoChartOptions.geoData;
+            legendProps = { ...legendProps, ...this.geoChartOptions.geoData };
+            if (size) {
+                const sizeLegendOptions = {
+                    sizeData: this.geoChartOptions.sizeValues || [],
+                    sizeFormat: getFormatFromExecutionResponse(size.index, executionResponse),
+                };
+                legendProps = { ...legendProps, ...sizeLegendOptions };
+            }
+            if (segmentBy) {
+                const segmentByLegendOptions = {
+                    segmentByData: this.geoChartOptions.series || [],
+                    legendItemsEnabled: this.state.legendIndexEnabled,
+                    onItemClick: this.onLegendItemClick,
+                };
+                legendProps = { ...legendProps, ...segmentByLegendOptions };
+            }
+        }
```

Comment:
Please move this logic to `getLegendProps`

Reply:
fixed

Reply:
It could be 
```
if (!this.geoChartOptions) { return legendProps; } 
const { size, segmentBy } = this.geoChartOptions.geoData;
            legendProps = { ...legendProps, geoData: { ...this.geoChartOptions.geoData } };
            if (size) {
                const sizeLegendOptions = {
                    sizeData: this.geoChartOptions.sizeData || [],
                    sizeFormat: getFormatFromExecutionResponse(size.index, executionResponse),
                };
                legendProps = { ...legendProps, ...sizeLegendOptions };
            }
            if (segmentBy) {
                const segmentByLegendOptions = {
                    segmentByData: this.geoChartOptions.segmentByData || [],
                    onItemClick: this.onLegendItemClick,
                };
                legendProps = { ...legendProps, ...segmentByLegendOptions };
            }
return legendProps;

```

Reply:
I mean we should avoid nested if as much as possible to improve readability

Reply:
fixed